File: binary-tree/python/binaryTree.py
import sys
from os.path import abspath

'''
# Work Node directory
node_dir = abspath('C:/github-projects/data-structures/shared/')
# Home Node directory
#node_dir = r'D:\github-projects\data-structures\shared'
sys.path.insert(0, node_dir)
from node import Node
'''

DEBUG = True

class Node:
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

    def getChildren(self):
        children = []
        if self.left:
            children.append(self.left)
        if self.right:
            children.append(self.right)
        return children

class BinaryTree:

    def __init__(self):
        self.root = None
        self.size = 0
        self.height = 0

    def insert(self, n, current = None, height=1):
        # Special case: insert into an empty tree
        if self.size == 0:
            self.root = n
            self.size += 1
            return

        # By default, start inserts at the root node
        if current is None:
            current = self.root

        # Insert left
        if (n.data <= current.data):
            if (current.left is None):
                current.left = n
                n.parent = current
                if (height > self.height):
                    self.height = height
                self.size += 1
            else:
                height += 1
                self.insert(n, current.left, height)

        # Insert right
        else:
            if (current.right is None):
                current.right = n
                n.parent = current
                if (height > self.height):
                    self.height = height
                self.size += 1
            else:
                height += 1
                self.insert(n, current.right, height)

    def getNodesDFS(self, n=None, nodes=[]):
        if not self.root:
            print("Error in getNodesDFS(): tree is empty")
        # Handle initial insert
        if not n:
            n = self.root
            nodes.append(n)
        # Standard insert
        else:
            nodes.append(n)
        # Recurse down the tree
        if n.left:
            self.getNodesDFS(n.left, nodes)
        if n.right:
            self.getNodesDFS(n.right, nodes)
        return nodes

    def getNodesBFS(self):
        nodes = []
        queue = []
        children = []
        if not self.root:
            print("Error in getNodesBFS: tree is empty")
        # Handle the root node
        nodes.append(self.root)
        queue.append(self.root)

        # Get all child nodes at each level
        while len(queue) > 0:
            children = queue[0].getChildren()
            for child in children:
                nodes.append(child)
                queue.append(child)
            queue.remove(queue[0])
        return nodes
        

    # Traversal functions
    def printDFS(self, n=None, level=0):
        if not self.root:
            print("Error in printDFS(): cannot print empty tree")
            # Base case: node is the root
        if not n:
            print("DFS: The tree has {} elements and is height {}".\
                format(self.size, self.height))   
            n = self.root
            print("Root: {:4}".format(n.data))
        else:
            print("Node: {:4}\tLevel: {}".format(n.data, level))
        if n.left:
            level += 1
            self.printDFS(n.left, level)
        if n.right:
            level -= 1
            self.printDFS(n.right, level)

    def printBFS(self, n=None, children=[], height=0):
        if not self.root:
            print("Error in printBFS(): cannot print empty tree")
        # Base case: node is the root
        if not n:
            children = self.root.getChildren()
            print("BFS: The tree has {} elements and is height {}".\
                format(self.size, self.height))
            n = self.root
            print("Root: {:4}\tHeight: {}".format(n.data, height))
        
        for child in children:
            print("Node: {:4}\tHeight: {}".format(child.data, height))


# Helper functions
def dPrint(message):
    if DEBUG: print("\tDebug: {}".format(message))

def nodesToString(nodes):
    ret = ""
    for node in nodes:
        ret += str(node.data) + " "
    #ret += "\n"
    return ret

File: binary-tree/c++/binaryTree.cpp
#include <iostream>
#include <iomanip>
#include <algorithm>
#include <vector>
#include <cmath>
#include <string>

#include "binaryTree.h"

BinaryTree::BinaryTree() {

    // Assign root node
    this->root = nullptr;
    // Set height and size
    this->height = 0;
    this->size = 0;
}

/* ***** Binary Tree functions ***** */
int BinaryTree::insert(struct Node *newNode, struct Node *parent, int height) {

    // Insert into empty tree
    if (this->size == 0) {
        this->root = newNode;
        this->size++;
        return 0;
    }

    // If not empty, ensure node ptr exists
    if (!parent) {
        return -1;
    }

    // Insert left
    else if (newNode->data <= parent->data) {
        // Insert left if no left child
        if (!(parent->left)) {
            // Set parent left and new node's parent
            parent->left = newNode;
            newNode->parent = parent;

            // Update height if changed
            if (height > this->height) {
                this->height = height;
            }
            
            // Update size
            this->size++;
            return 0;
        }
        // Otherwise recursively insert left
        else {
            return(insert(newNode, parent->left, height+1));
        }
    }

    // Insert right
    else {
        // Insert right if no right child
        if (!(parent->right)) {
            // Set parent left and new node's parent
            parent->right = newNode;
            newNode->parent = parent;

            // Update height if changed
            if (height > this->height) {
                this->height = height;
            }
            
            // Update size
            this->size++;
            return 0;
        }
        // Otherwise recursively insert right
        else {
            return(insert(newNode, parent->right, height+1));
        }
    }
}

/* ***** Accessors ***** */
int BinaryTree::getHeight() {
    return this->height;
}

long int BinaryTree::getSize() {
    return this->size;
}

struct Node *BinaryTree::getRoot() {
    return this->root;
}

/* Print Functions */
void BinaryTree::print() {
    if (this->size == 0) {
        std::cerr << "Error: cannot print empty tree\n";
        return;
    }

    // Height used to format printing
    int width = pow(2, height+1)-1;
    int offset = this->height;
    std::vector<struct Node*> parents{};
    std::vector<int> parentIndices{};
    std::vector<struct Node*> children{};
    std::vector<int> indices{};
    // Initialize vectors
    for (int i = 0; i < width; i++) {
        indices.push_back(0);
    }
    //bool root = true;

    // Print root first
    parents.push_back(this->getRoot());
    indices[width/2] = 1;
    parentIndices.push_back(width/2);
    printLine(parents, indices);
    
    while (parents.size() != 0) {
        // Zero out indices
        std::fill(indices.begin(), indices.end(), 0);

        // Set indices to be printed and collect children
        for (unsigned int i = 0; i < parents.size(); i++) {
            struct Node *tmp = parents[i];
            
            // Add children's children
            if (tmp->left) {
                children.push_back(tmp->left);
                indices[parentIndices.back()-offset] = 1;
            }
            if (tmp->right) {
                children.push_back(tmp->right);
                indices[parentIndices.back()+offset] = 1;
            }

            // Remove parent
            //parents.pop_back();
            parentIndices.pop_back();
        }
        printLine(children, indices);
        // Reduce offset
        offset--;
        // Previous children become next parents
        parents = children;
        // Indices become new parentIndices
        parentIndices.clear();
        getIndexLocations(indices, parentIndices);
        // Clear the old children
        children.clear();
        //std::reverse(parents.begin(), parents.end());
    }

}

void printLine(std::vector<struct Node *> nodes, std::vector<int> indices) {
    std::string sep = "____";
    int nodeIndex = 0;
    for (auto i : indices) {
        if (i == 1) {
            std::cout << nodes[nodeIndex]->data;
            nodeIndex++;
        }
        else {
            std::cout << sep;
        }
    }
    std::cout << std::endl;
}

void getIndexLocations(std::vector<int> &indices, \
    std::vector<int> &parentIndices) {
        for (int i = (int)indices.size()-1; i > 0; i--) {
            if (indices[i] == 1) {
                parentIndices.push_back(i);
            }
        }
}